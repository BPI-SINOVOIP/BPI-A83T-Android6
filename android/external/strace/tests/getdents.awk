BEGIN {
	i = "[0-9]+"
	len = "[1-9]" i

	d_ino = "d_ino=" i
	d_off = "d_off=" i
	d_reclen = "d_reclen=" len
	d_name_1 = "d_name=\"\\.\""
	d_name_2 = "d_name=\"\\.\\.\""
	d_name_3 = "d_name=\"(A\\\\n){127}Z\""
	# Some older systems might not pass back d_type at all like Alpha.
	d_type_dir = "d_type=DT_(DIR|UNKNOWN)"
	d_type_reg = "d_type=DT_(REG|UNKNOWN)"

	dirent_1   = "\\{" d_ino ", " d_off ", " d_reclen ", " d_name_1 ", " d_type_dir "\\}"
	dirent_2   = "\\{" d_ino ", " d_off ", " d_reclen ", " d_name_2 ", " d_type_dir "\\}"
	dirent_3   = "\\{" d_ino ", " d_off ", " d_reclen ", " d_name_3 ", " d_type_reg "\\}"

	dirent64_1 = "\\{" d_ino ", " d_off ", " d_reclen ", " d_type_dir ", " d_name_1 "\\}"
	dirent64_2 = "\\{" d_ino ", " d_off ", " d_reclen ", " d_type_dir ", " d_name_2 "\\}"
	dirent64_3 = "\\{" d_ino ", " d_off ", " d_reclen ", " d_type_reg ", " d_name_3 "\\}"

	d_123 = dirent_1 " " dirent_2 " " dirent_3
	d_213 = dirent_2 " " dirent_1 " " dirent_3
	d_132 = dirent_1 " " dirent_3 " " dirent_2
	d_321 = dirent_3 " " dirent_2 " " dirent_1
	d_231 = dirent_2 " " dirent_3 " " dirent_1
	d_312 = dirent_3 " " dirent_1 " " dirent_2

	d64_123 = dirent64_1 " " dirent64_2 " " dirent64_3
	d64_213 = dirent64_2 " " dirent64_1 " " dirent64_3
	d64_132 = dirent64_1 " " dirent64_3 " " dirent64_2
	d64_321 = dirent64_3 " " dirent64_2 " " dirent64_1
	d64_231 = dirent64_2 " " dirent64_3 " " dirent64_1
	d64_312 = dirent64_3 " " dirent64_1 " " dirent64_2

	dents = "\\{(" d_123 "|" d_213 "|" d_132 "|" d_321 "|" d_231 "|" d_312 ")\\}"
	dents64 = "\\{(" d64_123 "|" d64_213 "|" d64_132 "|" d64_321 "|" d64_231 "|" d64_312 ")\\}"

	getdents   =   "getdents\\(" i ", " dents   ", " len "\\)"
	getdents64 = "getdents64\\(" i ", " dents64 ", " len "\\)"

	r[1] = "^(" getdents "|" getdents64 ") += " len "$"
	r[2] = "^getdents(64)?\\([0-9]+, \\{\\}, [1-9][0-9]+\\) += 0$"
	s[3] = "+++ exited with 0 +++"

	lines = 3
	fail = 0
}

@include "match.awk"
