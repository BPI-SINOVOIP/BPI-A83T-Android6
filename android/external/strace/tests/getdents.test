#!/bin/sh

# Check that getdents/getdents64 syscalls are traced properly.

. "${srcdir=.}/init.sh"

check_prog gawk
AWKPATH="$srcdir" gawk -f "$srcdir"/getdents.awk "$srcdir"/getdents.out ||
	framework_skip_ 'gawk does not work properly'

check_prog ls
check_prog mkdir
check_prog rm
check_prog seq
check_prog touch

dir="$LOG.dir"
rm -rf -- "$dir"
mkdir -- "$dir" ||
	framework_skip_ 'failed to create a directory'

touch -- "$dir/$(for i in $(seq 1 127); do echo A; done; echo Z)" ||
	framework_skip_ 'failed to create a file'

ls -- "$dir" > /dev/null ||
	framework_skip_ 'failed to list a directory'

run_strace -vegetdents,getdents64 ls -- "$dir" > /dev/null
match_awk

rm -rf -- "$dir"

exit 0
