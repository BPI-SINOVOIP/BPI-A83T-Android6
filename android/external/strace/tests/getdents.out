getdents(3, {{d_ino=123456789, d_off=1, d_reclen=24, d_name=".", d_type=DT_DIR} {d_ino=234567890, d_off=2, d_reclen=24, d_name="..", d_type=DT_DIR} {d_ino=345678901, d_off=3, d_reclen=280, d_name="A\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nA\nZ", d_type=DT_REG}}, 4096) = 328
getdents(3, {}, 4096)                   = 0
+++ exited with 0 +++
